google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/DmlBatch.java

```diff
@@ -0,0 +1,190 @@
+ * called.
+ */
+class DmlBatch extends AbstractBaseUnitOfWork {
+  private final UnitOfWork transaction;
+  private final List<ParsedStatement> statements = new ArrayList<>();
+  private UnitOfWorkState state = UnitOfWorkState.STARTED;
+
+  static class Builder extends AbstractBaseUnitOfWork.Builder<Builder, DmlBatch> {
+    private UnitOfWork transaction;
+
+    private Builder() {}
+
+    Builder setTransaction(UnitOfWork transaction) {
+      Preconditions.checkNotNull(transaction);
+      this.transaction = transaction;
+      return this;
+    }
+
+    @Override
+    DmlBatch build() {
+      Preconditions.checkState(transaction != null, "No transaction specified");
+      return new DmlBatch(this);
+    }
+  }
+
+  static Builder newBuilder() {
+    return new Builder();
+  }
+
+  private DmlBatch(Builder builder) {
+    super(builder);
+    this.transaction = builder.transaction;
+  }
+
+  @Override
+  public Type getType() {
+    return Type.BATCH;
+  }
+
+  @Override
+  public UnitOfWorkState getState() {
+    return this.state;
+  }
+
+  @Override
+  public boolean isActive() {
+    return getState().isActive();
+  }
+
+  @Override
+  public boolean isReadOnly() {
+    return false;
+  }
+
+  @Override
+  public ResultSet executeQuery(ParsedStatement statement, AnalyzeMode analyzeMode,
+      QueryOption... options) {
+    throw SpannerExceptionFactory.newSpannerException(ErrorCode.FAILED_PRECONDITION,
+        "Executing queries is not allowed for DML batches.");
+  }
+
+  @Override
+  public Timestamp getReadTimestamp() {
+    throw SpannerExceptionFactory.newSpannerException(ErrorCode.FAILED_PRECONDITION,
+        "There is no read timestamp available for DML batches.");
+  }
+
+  @Override
+  public Timestamp getReadTimestampOrNull() {
+    return null;
+  }
+
+  @Override
+  public Timestamp getCommitTimestamp() {
+    throw SpannerExceptionFactory.newSpannerException(ErrorCode.FAILED_PRECONDITION,
+        "There is no commit timestamp available for DML batches.");
+  }
+
+  @Override
+  public Timestamp getCommitTimestampOrNull() {
+    return null;
+  }
+
+  @Override
+  public void executeDdl(ParsedStatement ddl) {
+    throw SpannerExceptionFactory.newSpannerException(ErrorCode.FAILED_PRECONDITION,
+        "Executing DDL statements is not allowed for DML batches.");
+  }
+
+  @Override
+  public long executeUpdate(ParsedStatement update) {
+    ConnectionPreconditions.checkState(state == UnitOfWorkState.STARTED,
+        "The batch is no longer active and cannot be used for further statements");
+    Preconditions.checkArgument(update.getType() == StatementType.UPDATE,
+        "Only DML statements are allowed. \"" + update.getSqlWithoutComments()
+            + "\" is not a DML-statement.");
+    statements.add(update);
+    return -1L;
+  }
+
+  @Override
+  public long[] executeBatchUpdate(Iterable<ParsedStatement> updates) {
+    throw SpannerExceptionFactory.newSpannerException(ErrorCode.FAILED_PRECONDITION,
```

Comment:
This is surprising to me from an API perspective. We can call executeUpdate, then runBatch(), but I would have expected this to be somewhat of a convenience method?

Reply:
I understand the confusion, but `DmlBatch` should be seen as a kind of wrapper around the fact that both standard SQL and JDBC do not have any `executeBatchUpdate(Iterable)` equivalent. Instead, both work by declaring the start of a batch, then add statements to the batch, and then execute the batch.

In case of SQL:

```sql
START BATCH DML;
UPDATE FOO SET BAR=1 WHERE ID=1;
UPDATE FOO SET BAR=2 WHERE ID=2;
RUN BATCH;
```

The following (or similar) is not possible in standard SQL:

```sql
EXECUTE('UPDATE FOO SET BAR=1 WHERE ID=1', 'UPDATE FOO SET BAR=2 WHERE ID=2')
```

Much the same applies to JDBC:

```java
Statement statement = createStatemet(...);
statement.addBatch("UPDATE FOO SET BAR=1 WHERE ID=1");
statement.addBatch("UPDATE FOO SET BAR=2 WHERE ID=2");
statement.executeBatch();
```

Allowing it as a convenience method could maybe have been an option if it was a public interface, but as this is an internal interface my opinion is that we should be strict when it comes to correct usage.

Reply:
Thanks, that makes sense! Also didn't realize this wasn't a public API 😄